Firmware/AVR/AVR_SENSOR/lib/ir_sensor/ir_sensor.c

```diff
@@ -0,0 +1,185 @@
+    while (ADCSRA & (1 << ADSC) ); // wait till conversion complete 
+    ir_data.ir_front_1_data = ADCH;
+
+    ADMUX = IR_FRONT_2_MUX;
+    ADCSRA |= (1 << ADSC);         // start ADC measurement
+    while (ADCSRA & (1 << ADSC) ); // wait till conversion complete 
+    ir_data.ir_front_2_data = ADCH;
+
+    // Right IR Sensors
+    ADMUX = IR_RIGHT_1_MUX;
+    ADCSRA |= (1 << ADSC);         // start ADC measurement
+    while (ADCSRA & (1 << ADSC) ); // wait till conversion complete 
+    ir_data.ir_right_1_data = ADCH;
+
+    ADMUX = IR_RIGHT_2_MUX;
+    ADCSRA |= (1 << ADSC);         // start ADC measurement
+    while (ADCSRA & (1 << ADSC) ); // wait till conversion complete 
+    ir_data.ir_right_2_data = ADCH;
+
+    // Left IR Sensors
+    ADMUX = IR_LEFT_1_MUX;
+    ADCSRA |= (1 << ADSC);         // start ADC measurement
+    while (ADCSRA & (1 << ADSC) ); // wait till conversion complete 
+    ir_data.ir_left_1_data = ADCH;
+
+    ADMUX = IR_LEFT_2_MUX;
+    ADCSRA |= (1 << ADSC);         // start ADC measurement
+    while (ADCSRA & (1 << ADSC) ); // wait till conversion complete 
+    ir_data.ir_left_2_data = ADCH;    
+}
+
+ir_data_S ir_sensor_get(void)
+{
+    return ir_data;
+}
+
+ir_data_S ir_sensor_retrieve(void)
+{
+    ir_data_S ir_temp_data;
+
+    // Front IR sensors
+    ADMUX = IR_FRONT_1_MUX;
+    ADCSRA |= (1 << ADSC);         // start ADC measurement
+    while (ADCSRA & (1 << ADSC) ); // wait till conversion complete 
+    ir_temp_data.ir_front_1_data = ADCH;
+
+    ADMUX = IR_FRONT_2_MUX;
+    ADCSRA |= (1 << ADSC);         // start ADC measurement
+    while (ADCSRA & (1 << ADSC) ); // wait till conversion complete 
+    ir_temp_data.ir_front_2_data = ADCH;
+
+    // Right IR Sensors
+    ADMUX = IR_RIGHT_1_MUX;
+    ADCSRA |= (1 << ADSC);         // start ADC measurement
+    while (ADCSRA & (1 << ADSC) ); // wait till conversion complete 
+    ir_temp_data.ir_right_1_data = ADCH;
+
+    ADMUX = IR_RIGHT_2_MUX;
+    ADCSRA |= (1 << ADSC);         // start ADC measurement
+    while (ADCSRA & (1 << ADSC) ); // wait till conversion complete 
+    ir_temp_data.ir_right_2_data = ADCH;
+
+    // Left IR Sensors
+    ADMUX = IR_LEFT_1_MUX;
+    ADCSRA |= (1 << ADSC);         // start ADC measurement
+    while (ADCSRA & (1 << ADSC) ); // wait till conversion complete 
+    ir_temp_data.ir_left_1_data = ADCH;
+
+    ADMUX = IR_LEFT_2_MUX;
+    ADCSRA |= (1 << ADSC);         // start ADC measurement
+    while (ADCSRA & (1 << ADSC) ); // wait till conversion complete 
+    ir_temp_data.ir_left_2_data = ADCH;
+
+    return ir_temp_data;
+}
+
+void ir_test_code(void) 
+{
+  
+    ir_attiny_init();
+
+    ir_data_S ir_sensor_data;
+
+    while(1)
+    {
+        ir_sensor_data = ir_sensor_retrieve();
+
+        char my_string[7];
+
+        sprintf(my_string, "IR_DATA = %d\n", ir_sensor_data.ir_left_2_data);
+        UART_tx_str(my_string);
+
+        _delay_ms(500/8);
```

Comment:
This would not necessary guarantee the update frequency, we might need to actively compute the delta time to delay: _delay_ms(delta_ticks)